### PR TITLE
Improve support for runtime bean registration

### PR DESCRIPTION
**Pull request by graemerocher**
Currently it isn't possible to register a bean at runtime that is lazily created by the container nor is it possible to register beans in different scopes. This is needed for improvements to Spring bean import. I have submitted against 3.5.x because if it doesn't go into that branch then we have a chicken and egg problem with the 3.6 release in terms of getting improvements into that release.

### CODE REVIEW

File: inject/src/main/java/io/micronaut/context/BeanDefinitionReferenceRegistry.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2017-2020 original authors
```
Comment (burtbeckwith) [inline inject/src/main/java/io/micronaut/context/BeanDefinitionReferenceRegistry.java]: 2022

**Review by burtbeckwith — APPROVED** [review]

**Review by dstepanov — APPROVED** [review]

**Comment by dstepanov** [comment]
I wonder if it still makes sense to have the separate concept of bean references and bean definitions. I feel like we should just stick with BeanDefinitions in v4 and reduce the number of extra classes generated.

**Comment by graemerocher** [comment]
Perhaps not, we can revisit that in 4.0

**Review by sdelamo — APPROVED** [review]

**Comment by graemerocher** [comment]
hold off merging have a few changes to add

**Comment by sdelamo** [comment]
I am not sure about this going into 3.5.x. I think we should target 3.6.x. Once 3.6.0 is out we can release a next minor of Micronaut Spring referencing core 3.6.0. thoughts @graemerocher ?

**Comment by sdelamo** [comment]
> hold off merging have a few changes to add

I have set the PR to draft.

**Review by sdelamo — APPROVED** [review]
Looks good to me. I have created a PR with small changes 

https://github.com/micronaut-projects/micronaut-core/pull/7606

**Pull request merged**